pyls/rpc_manager.py

```diff
@@ -0,0 +1,171 @@
+        self._message_handler = message_handler
+        self._shutdown = False
+        self._sent_requests = {}
+        self._received_requests = {}
+        self._executor_service = ThreadPoolExecutor()
+
+    def start(self):
+        """Start reading JSONRPC messages off of rx"""
+        self.consume_requests()
+
+    def shutdown(self):
+        """Set flag to ignore all non exit messages"""
+        self._shutdown = True
+
+    def exit(self):
+        """Stop listening for new message"""
+        self._executor_service.shutdown()
+        self._message_manager.close()
+
+    def call(self, method, params=None):
+        """Send a JSONRPC message with an expected response.
+
+        Args:
+            method (str): The method name of the message to send
+            params (dict): The payload of the message
+
+        Returns:
+            Future that will resolve once a response has been recieved
+
+        """
+        log.debug('Calling %s %s', method, params)
+        request = JSONRPC20Request(_id=uuid1().int, method=method, params=params)
+        request_future = Future()
+        self._sent_requests[request._id] = request_future
+        self._message_manager.write_message(request.data)
+        return request_future
+
+    def notify(self, method, params=None):
+        """Send a JSONRPC notification.
+
+         Args:
+             method (str): The method name of the notification to send
+             params (dict): The payload of the notification
+         """
+        log.debug('Notify %s %s', method, params)
+        notification = JSONRPC20Request(method=method, params=params)
+        self._message_manager.write_message(notification.data)
+
+    def cancel(self, request_id):
+        """Cancel pending request handler.
+
+        Args:
+            request_id (string | number): The id of the original request
+
+        Note:
+            Request will only be cancelled if it has not begun execution.
+        """
+        log.debug('Cancel request %d', request_id)
+        try:
+            self._received_requests[request_id].cancel()
+        except KeyError:
+            log.error('Received cancel for finished/nonexistent request %d', request_id)
+
+    def consume_requests(self):
+        """ Infinite loop watching for messages from the client."""
+        for message in self._message_manager.get_messages():
+            if isinstance(message, JSONRPCBaseResponse):
+                self._handle_response(message)
+            else:
+                self._handle_request(message)
+
+    def _handle_request(self, request):
+        """Execute corresponding handler for the recieved request
+
+        Args:
+            request (JSONRPCBaseRequest): request to act upon
+
+        Note:
+            requests are handled asynchronously if the handler returns a callable, otherwise they are handle
+            synchronously by the main thread
+        """
+        if self._shutdown and request.method != 'exit':
```

Comment:
What is this for? Why does the server know about a 'exit' method?

Reply:
I suppose the language server can `drop` the message within the handler, but it complicates the logic since this method now has to know how to differentiate between an empty response and a response that should be dropped